File: app/src/main/java/com/cyl/musiclake/view/lyric/FloatLyricView.java
package com.cyl.musiclake.view.lyric;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import com.cyl.musiclake.R;
import com.cyl.musiclake.common.NavigationHelper;
import com.cyl.musiclake.player.MusicPlayerService;
import com.cyl.musiclake.utils.LogUtil;
import com.cyl.musiclake.utils.SPUtils;
import com.cyl.musiclake.utils.ToastUtils;
import com.rtugeek.android.colorseekbar.ColorSeekBar;

import net.steamcrafted.materialiconlib.MaterialDrawableBuilder;
import net.steamcrafted.materialiconlib.MaterialIconView;

import java.lang.reflect.Field;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;

/**
 * 桌面歌词View
 */
public class FloatLyricView extends FrameLayout implements View.OnClickListener {

    /**
     * 记录小悬浮窗的宽度
     */
    public int viewWidth;

    /**
     * 记录小悬浮窗的高度
     */
    public int viewHeight;

    /**
     * 记录系统状态栏的高度
     */
    private static int statusBarHeight;

    /**
     * 用于更新小悬浮窗的位置
     */
    private WindowManager windowManager;

    /**
     * 小悬浮窗的参数
     */
    private WindowManager.LayoutParams mParams;

    /**
     * 记录当前手指位置在屏幕上的横坐标值
     */
    private float xInScreen;

    /**
     * 记录当前手指位置在屏幕上的纵坐标值
     */
    private float yInScreen;

    /**
     * 记录手指按下时在屏幕上的横坐标的值
     */
    private float xDownInScreen;

    /**
     * 记录手指按下时在屏幕上的纵坐标的值
     */
    private float yDownInScreen;

    /**
     * 记录手指按下时在小悬浮窗的View上的横坐标的值
     */
    private float xInView;

    /**
     * 记录手指按下时在小悬浮窗的View上的纵坐标的值
     */
    private float yInView;
    private float mFontSize;
    private int mFontColor;
    private boolean mMovement;
    private boolean isHiddenSettings;

    public LyricTextView mLyricText;
    public TextView mTitle;
    public SeekBar mSizeSeekBar;
    public ColorSeekBar mColorSeekBar;
    private MaterialIconView mLockButton, mPreButton, mNextButton, mPlayButton, mSettingsButton;
    private ImageButton mCloseButton;
    private ImageButton mMusicButton;
    private LinearLayout mSettingLinearLayout;
    private LinearLayout mRelLyricView;
    private LinearLayout mLinLyricView;
    private FrameLayout mFrameBackground;
    private View mRootView;

    public FloatLyricView(Context context) {
        super(context);
        windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        mRootView = LayoutInflater.from(context).inflate(R.layout.float_lyric_view, this);
        FrameLayout view = findViewById(R.id.small_window_layout);
        viewWidth = view.getLayoutParams().width;
        viewHeight = view.getLayoutParams().height;
        mMovement = true;
        isHiddenSettings = true;

        mTitle = findViewById(R.id.music_title);
        mSizeSeekBar = findViewById(R.id.sb_size);
        mColorSeekBar = findViewById(R.id.sb_color);
        mLyricText = findViewById(R.id.lyric);
        mCloseButton = findViewById(R.id.btn_close);
        mMusicButton = findViewById(R.id.music_app);
        mLockButton = findViewById(R.id.btn_lock);
        mPreButton = findViewById(R.id.btn_previous);
        mPlayButton = findViewById(R.id.btn_play);
        mNextButton = findViewById(R.id.btn_next);
        mSettingsButton = findViewById(R.id.btn_settings);
        mSettingLinearLayout = findViewById(R.id.ll_settings);
        mRelLyricView = findViewById(R.id.rl_layout);
        mLinLyricView = findViewById(R.id.ll_layout);
        mFrameBackground = findViewById(R.id.small_bg);

        mCloseButton.setOnClickListener(this);
        mMusicButton.setOnClickListener(this);
        mLockButton.setOnClickListener(this);
        mPreButton.setOnClickListener(this);
        mPlayButton.setOnClickListener(this);
        mNextButton.setOnClickListener(this);
        mSettingsButton.setOnClickListener(this);

        mFontSize = SPUtils.getFontSize();
        mLyricText.setFontSizeScale(mFontSize);
        mSizeSeekBar.setProgress((int) mFontSize);

        mFontColor = SPUtils.getFontColor();
        mLyricText.setFontColorScale(mFontColor);
        mColorSeekBar.setColorBarPosition(mFontColor);

        setPlayStatus(MusicPlayerService.getInstance().isPlaying());

        mSizeSeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                LogUtil.e("TEST", progress + "---" + fromUser);
                mLyricText.setFontSizeScale(progress);
                SPUtils.saveFontSize(progress);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
        mColorSeekBar.setOnColorChangeListener((colorBarPosition, alphaBarPosition, color) -> {
            mLyricText.setFontColorScale(color);
            SPUtils.saveFontColor(color);
        });

    }


    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                // 手指按下时记录必要数据,纵坐标的值都需要减去状态栏高度
                xInView = event.getX();
                yInView = event.getY();
                xDownInScreen = event.getRawX();
                yDownInScreen = event.getRawY() - getStatusBarHeight();
                xInScreen = event.getRawX();
                yInScreen = event.getRawY() - getStatusBarHeight();
                break;
            case MotionEvent.ACTION_MOVE:
                xInScreen = event.getRawX();
                yInScreen = event.getRawY() - getStatusBarHeight();
                // 手指移动的时候更新小悬浮窗的位置
                updateViewPosition();
                break;
            case MotionEvent.ACTION_UP:
                // 如果手指离开屏幕时，xDownInScreen和xInScreen相等，且yDownInScreen和yInScreen相等，则视为触发了单击事件。
                if (xDownInScreen == xInScreen && yDownInScreen == yInScreen && mMovement) {
                    toggleLyricView();
                }
                break;
            default:
                break;
        }
        return super.onTouchEvent(event);
    }

    /**
     * 将小悬浮窗的参数传入，用于更新小悬浮窗的位置。
     *
     * @param params 小悬浮窗的参数
     */
    public void setParams(WindowManager.LayoutParams params) {
        mParams = params;
    }

    /**
     * 更新小悬浮窗在屏幕中的位置。
     */
    private void updateViewPosition() {
        if (!mMovement) return;
        mParams.x = (int) (xInScreen - xInView);
        mParams.y = (int) (yInScreen - yInView);
        windowManager.updateViewLayout(this, mParams);
    }

    /**
     * toggle背景
     */
    private void toggleLyricView() {
        if (mRootView != null) {
            if (mRelLyricView.getVisibility() == View.INVISIBLE) {
                mRelLyricView.setVisibility(View.VISIBLE);
                mLinLyricView.setVisibility(View.VISIBLE);
                mFrameBackground.setVisibility(View.VISIBLE);
            } else {
                if (!isHiddenSettings) {
                    isHiddenSettings = true;
                    updateSettingStatus(isHiddenSettings);
                }
                mLinLyricView.setVisibility(View.INVISIBLE);
                mRelLyricView.setVisibility(View.INVISIBLE);
                mFrameBackground.setVisibility(View.INVISIBLE);
            }
        }
    }


    /**
     * 用于获取状态栏的高度。
     *
     * @return 返回状态栏高度的像素值。
     */
    private int getStatusBarHeight() {
        if (statusBarHeight == 0) {
            try {
                Class<?> c = Class.forName("com.android.internal.R$dimen");
                Object o = c.newInstance();
                Field field = c.getField("status_bar_height");
                int x = (Integer) field.get(o);
                statusBarHeight = getResources().getDimensionPixelSize(x);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        return statusBarHeight;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.music_app:
                Intent intent = NavigationHelper.INSTANCE.getNowPlayingIntent(getContext());
                intent.setFlags(FLAG_ACTIVITY_NEW_TASK);
                getContext().startActivity(intent);
                break;
            case R.id.btn_close:
                MusicPlayerService.getInstance().showDesktopLyric(false);
                break;
            case R.id.btn_lock:
                mMovement = !mMovement;
                if (mMovement) {
                    mLockButton.setIcon(MaterialDrawableBuilder.IconValue.LOCK_OPEN);
                } else {
                    toggleLyricView();
                    ToastUtils.show("桌面歌词已锁定,请前往通知栏或者设置解锁");
                    mLockButton.setIcon(MaterialDrawableBuilder.IconValue.LOCK);
                }
                break;
            case R.id.btn_previous:
                MusicPlayerService.getInstance().prev();
                break;
            case R.id.btn_play:
                MusicPlayerService.getInstance().playPause();
                setPlayStatus(MusicPlayerService.getInstance().isPlaying());
                break;
            case R.id.btn_next:
                MusicPlayerService.getInstance().next(false);
                break;
            case R.id.btn_settings:
                isHiddenSettings = !isHiddenSettings;
                updateSettingStatus(isHiddenSettings);
                break;
        }
    }

    public void setPlayStatus(boolean isPlaying) {
        if (isPlaying) {
            mPlayButton.setIcon(MaterialDrawableBuilder.IconValue.PAUSE);
        } else {
            mPlayButton.setIcon(MaterialDrawableBuilder.IconValue.PLAY);
        }
    }

    public void updateSettingStatus(boolean isHidden) {
        if (isHidden) {
            mSettingLinearLayout.setVisibility(GONE);
        } else {
            mSettingLinearLayout.setVisibility(VISIBLE);
        }
    }

}

File: app/src/main/java/com/cyl/musiclake/ui/widget/lyric/FloatLyricView.kt
package com.cyl.musiclake.ui.widget.lyric

import android.content.Context
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.*
import com.cyl.musiclake.MusicApp
import com.cyl.musiclake.R
import com.cyl.musiclake.common.NavigationHelper
import com.cyl.musiclake.player.MusicPlayerService
import com.cyl.musiclake.player.UnLockNotify
import com.cyl.musiclake.utils.LogUtil
import com.cyl.musiclake.utils.SPUtils
import com.cyl.musiclake.utils.ToastUtils
import com.rtugeek.android.colorseekbar.ColorSeekBar
import net.steamcrafted.materialiconlib.MaterialDrawableBuilder
import net.steamcrafted.materialiconlib.MaterialIconView

/**
 * 桌面歌词View
 */
class FloatLyricView(context: Context) : FrameLayout(context), View.OnClickListener {

    /**
     * 记录小悬浮窗的宽度
     */
    var viewWidth: Int = 0

    /**
     * 记录小悬浮窗的高度
     */
    var viewHeight: Int = 0

    /**
     * 用于更新小悬浮窗的位置
     */
    private val windowManager: WindowManager

    /**
     * 小悬浮窗的参数
     */
    private var mParams: WindowManager.LayoutParams? = null

    /**
     * 记录当前手指位置在屏幕上的横坐标值
     */
    private var xInScreen: Float = 0.toFloat()

    /**
     * 记录当前手指位置在屏幕上的纵坐标值
     */
    private var yInScreen: Float = 0.toFloat()

    /**
     * 记录手指按下时在屏幕上的横坐标的值
     */
    private var xDownInScreen: Float = 0.toFloat()

    /**
     * 记录手指按下时在屏幕上的纵坐标的值
     */
    private var yDownInScreen: Float = 0.toFloat()

    /**
     * 记录手指按下时在小悬浮窗的View上的横坐标的值
     */
    private var xInView: Float = 0.toFloat()

    /**
     * 记录手指按下时在小悬浮窗的View上的纵坐标的值
     */
    private var yInView: Float = 0.toFloat()
    private val mFontSize: Float
    private val mFontColor: Int
    private var mMovement: Boolean = false
    private var isHiddenSettings: Boolean = false


    private val mRootView by lazy { LayoutInflater.from(context).inflate(R.layout.float_lyric_view, this) }
    private val mPreButton by lazy { mRootView?.findViewById<MaterialIconView>(R.id.btn_previous) }
    private val mPlayButton by lazy { mRootView?.findViewById<MaterialIconView>(R.id.btn_play) }
    private val mNextButton by lazy { mRootView?.findViewById<MaterialIconView>(R.id.btn_next) }
    private val mLockButton by lazy { mRootView?.findViewById<MaterialIconView>(R.id.btn_lock) }
    private val mSettingsButton by lazy { mRootView?.findViewById<MaterialIconView>(R.id.btn_settings) }
    private val mFrameBackground by lazy { mRootView?.findViewById<View>(R.id.small_bg) }
    private val mSettingLinearLayout by lazy { mRootView?.findViewById<View>(R.id.ll_settings) }
    private val mRelLyricView by lazy { mRootView?.findViewById<View>(R.id.rl_layout) }
    private val mLinLyricView by lazy { mRootView?.findViewById<View>(R.id.ll_layout) }

    private val mMusicButton by lazy { mRootView?.findViewById<ImageButton>(R.id.music_app) }
    private val mCloseButton by lazy { mRootView?.findViewById<ImageButton>(R.id.btn_close) }
    val mLyricText by lazy { mRootView?.findViewById<LyricTextView>(R.id.lyric) }
    val mTitle by lazy { mRootView?.findViewById<TextView>(R.id.music_title) }
    private val mSizeSeekBar by lazy { mRootView?.findViewById<SeekBar>(R.id.sb_size) }
    private val mColorSeekBar by lazy { mRootView?.findViewById<ColorSeekBar>(R.id.sb_color) }
    private val view by lazy { mRootView?.findViewById<FrameLayout>(R.id.small_window_layout) }

    private var mIsLock: Boolean = false
    private val mNotify: UnLockNotify

    init {
        windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        mNotify = UnLockNotify()

        viewWidth = view?.layoutParams?.width ?: 0
        viewHeight = view?.layoutParams?.height ?: 0
        mMovement = true
        isHiddenSettings = true


        mCloseButton?.setOnClickListener(this)
        mMusicButton?.setOnClickListener(this)
        mLockButton?.setOnClickListener(this)
        mPreButton?.setOnClickListener(this)
        mPlayButton?.setOnClickListener(this)
        mNextButton?.setOnClickListener(this)
        mSettingsButton?.setOnClickListener(this)

        mFontSize = SPUtils.getFontSize().toFloat()
        mLyricText?.setFontSizeScale(mFontSize)
        mSizeSeekBar?.progress = mFontSize.toInt()

        mIsLock = SPUtils.getAnyByKey(SPUtils.SP_KEY_FLOAT_LYRIC_LOCK, false)

        saveLock(mIsLock, false)

        mFontColor = SPUtils.getFontColor()
        mLyricText?.setFontColorScale(mFontColor)
        mColorSeekBar?.colorBarPosition = mFontColor

        setPlayStatus(MusicPlayerService.getInstance().isPlaying)

        mSizeSeekBar?.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                LogUtil.e("TEST", "$progress---$fromUser")
                mLyricText?.setFontSizeScale(progress.toFloat())
                SPUtils.saveFontSize(progress)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })
        mColorSeekBar?.setOnColorChangeListener { colorBarPosition, alphaBarPosition, color ->
            mLyricText?.setFontColorScale(color)
            SPUtils.saveFontColor(color)
        }

    }


    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                // 手指按下时记录必要数据,纵坐标的值都需要减去状态栏高度
                xInView = event.x
                yInView = event.y
                xDownInScreen = event.rawX
                yDownInScreen = event.rawY - getStatusBarHeight()
                xInScreen = event.rawX
                yInScreen = event.rawY - getStatusBarHeight()
            }
            MotionEvent.ACTION_MOVE -> {
                xInScreen = event.rawX
                yInScreen = event.rawY - getStatusBarHeight()
                // 手指移动的时候更新小悬浮窗的位置
                updateViewPosition()
            }
            MotionEvent.ACTION_UP ->
                // 如果手指离开屏幕时，xDownInScreen和xInScreen相等，且yDownInScreen和yInScreen相等，则视为触发了单击事件。
                if (xDownInScreen == xInScreen && yDownInScreen == yInScreen && mMovement) {
                    showLyricBackground()
                }
            else -> {
            }
        }
        return super.onTouchEvent(event)
    }

    /**
     * 将小悬浮窗的参数传入，用于更新小悬浮窗的位置。
     *
     * @param params 小悬浮窗的参数
     */
    fun setParams(params: WindowManager.LayoutParams) {
        mParams = params
    }

    /**
     * 更新小悬浮窗在屏幕中的位置。
     */
    private fun updateViewPosition() {
        if (!mMovement) return
        mParams!!.x = (xInScreen - xInView).toInt()
        mParams!!.y = (yInScreen - yInView).toInt()
        windowManager.updateViewLayout(this, mParams)
    }

    /**
     * 显示隐藏背景
     */
    private fun showLyricBackground() {
        LogUtil.d("FloatLyricView", "桌面歌词状态：mIsLock:$mIsLock")
        if (mRootView != null) {
            if (!mIsLock) {
                mRelLyricView?.visibility = View.VISIBLE
                mLinLyricView?.visibility = View.VISIBLE
                mFrameBackground?.visibility = View.VISIBLE
            } else {
                if (!isHiddenSettings) {
                    isHiddenSettings = true
                    updateSettingStatus(isHiddenSettings)
                }
                mLinLyricView?.visibility = View.INVISIBLE
                mRelLyricView?.visibility = View.INVISIBLE
                mFrameBackground?.visibility = View.INVISIBLE
            }
        }
    }


    /**
     * 用于获取状态栏的高度。
     *
     * @return 返回状态栏高度的像素值。
     */
    private fun getStatusBarHeight(): Int {
        if (statusBarHeight == 0) {
            try {
                val c = Class.forName("com.android.internal.R\$dimen")
                val o = c.newInstance()
                val field = c.getField("status_bar_height")
                val x = field.get(o) as Int
                statusBarHeight = resources.getDimensionPixelSize(x)
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
        return statusBarHeight
    }


    override fun onClick(v: View) {
        when (v.id) {
            R.id.music_app -> {
                val intent = NavigationHelper.getNowPlayingIntent(context)
                intent.flags = FLAG_ACTIVITY_NEW_TASK
                context.startActivity(intent)
            }
            R.id.btn_close -> MusicPlayerService.getInstance().showDesktopLyric(false)
            R.id.btn_lock -> {
                mMovement = !mMovement
                if (mMovement) {
                    mLockButton?.setIcon(MaterialDrawableBuilder.IconValue.LOCK_OPEN)
                } else {
                    saveLock(true, true)
                }
            }
            R.id.btn_previous -> MusicPlayerService.getInstance().prev()
            R.id.btn_play -> {
                MusicPlayerService.getInstance().playPause()
                setPlayStatus(MusicPlayerService.getInstance().isPlaying)
            }
            R.id.btn_next -> MusicPlayerService.getInstance().next(false)
            R.id.btn_settings -> {
                isHiddenSettings = !isHiddenSettings
                updateSettingStatus(isHiddenSettings)
            }
        }
    }

    fun setPlayStatus(isPlaying: Boolean) {
        if (isPlaying) {
            mPlayButton?.setIcon(MaterialDrawableBuilder.IconValue.PAUSE)
        } else {
            mPlayButton?.setIcon(MaterialDrawableBuilder.IconValue.PLAY)
        }
    }

    fun updateSettingStatus(isHidden: Boolean) {
        if (isHidden) {
            mSettingLinearLayout?.visibility = View.GONE
        } else {
            mSettingLinearLayout?.visibility = View.VISIBLE
        }
    }

    /**
     * 更新歌词锁定状态
     * @param lock 锁定
     * @param toast 是否弹出提示
     */
    fun saveLock(lock: Boolean, toast: Boolean) {
        mIsLock = lock
        LogUtil.d("FloatLyricView", "桌面歌词状态：mIsLock:$mIsLock$lock")
        SPUtils.putAnyCommit(SPUtils.SP_KEY_FLOAT_LYRIC_LOCK, mIsLock)
        if (toast) {
            ToastUtils.show(MusicApp.getAppContext(), if (!mIsLock) R.string.float_unlock else R.string.float_lock)
        }
        if (layoutParams != null) {
            val params = layoutParams as WindowManager.LayoutParams
            if (lock) {
                //锁定后点击通知栏解锁
                mNotify.notifyToUnlock()
                params.flags = (WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                        or WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
                mLockButton?.setIcon(MaterialDrawableBuilder.IconValue.LOCK)
            } else {
                mMovement = true
                mNotify.cancel()
                params.flags = WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                mLockButton?.setIcon(MaterialDrawableBuilder.IconValue.LOCK_OPEN)
            }
            showLyricBackground()
            windowManager.updateViewLayout(this, params)
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        mNotify.cancel()
    }

    companion object {

        /**
         * 记录系统状态栏的高度
         */
        private var statusBarHeight: Int = 0
    }

}